runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/ConnectionIdleMonitor.kt

```diff
@@ -0,0 +1,123 @@
+            logger.trace { "Cancel monitoring for $connection" }
+
+            // Use `runBlocking` because this _must_ finish before OkHttp goes to use the connection
+            val cancelTime = measureTime {
+                runBlocking(context) { monitor.cancelAndJoin() }
+            }
+
+            logger.trace { "Monitoring canceled for $connection in $cancelTime" }
+        }
+    }
+
+    override fun connectionReleased(connection: Connection, call: Call) {
+        val connId = System.identityHashCode(connection)
+        val context = call.callContext()
+        val scope = CoroutineScope(context)
+        val monitor = scope.launch(CoroutineName("okhttp-conn-monitor-for-$connId")) {
+            doMonitor(connection)
+        }
+        context.logger<ConnectionIdleMonitor>().trace { "Launched coroutine $monitor to monitor $connection" }
+
+        // Non-locking map access is okay here because this code will only execute synchronously as part of a
+        // `connectionReleased` event and will be complete before any future `connectionAcquired` event could fire for
+        // the same connection.
+        monitors[connection] = monitor
+    }
+
+    private suspend fun doMonitor(conn: Connection) {
+        val logger = coroutineContext.logger<ConnectionIdleMonitor>()
+
+        val socket = conn.socket()
+        val source = try {
+            socket.source()
+        } catch (_: SocketException) {
+            logger.trace { "Socket for $conn closed before monitoring started. Skipping polling loop." }
+            return
+        }.buffer().peek()
+
+        logger.trace { "Commence socket monitoring for $conn" }
+        var resetTimeout = true
+        val oldTimeout = socket.soTimeout
+
+        try {
+            socket.soTimeout = pollInterval.inWholeMilliseconds.toInt()
+
+            while (coroutineContext.isActive) {
+                try {
+                    logger.trace { "Polling socket for $conn" }
+                    source.readByte() // Blocking read; will take up to READ_TIMEOUT_MS to complete
+                } catch (_: SocketTimeoutException) {
+                    logger.trace { "Socket still alive for $conn" }
+                    // Socket is still alive
+                } catch (_: EOFException) {
+                    logger.trace { "Socket closed remotely for $conn" }
+                    socket.closeQuietly()
+                    resetTimeout = false
+                    return
+                }
+            }
+
+            logger.trace { "Monitoring coroutine has been cancelled. Ending polling loop." }
+        } catch (e: Throwable) {
+            logger.warn(e) { "Failed to poll $conn. Ending polling loop. Connection may be unstable now." }
+        } finally {
+            if (resetTimeout) {
+                logger.trace { "Attempting to reset soTimeout..." }
+                try {
+                    conn.socket().soTimeout = oldTimeout
```

Comment:
question: Is it possible for a monitored connection to be pulled and used by OkHttp before we reset the `soTimeout`?

Reply:
Do we have measurements for how long those health checks last and how long we take to reset the `soTimeout`? We might have a lot of overhead or maybe not.

Reply:
Nvm this comment, I was just misunderstanding something